File: 2023_8/baekjoon/17103.py
def sieve_of_eratosthenes(limit):
    primes = {i: True for i in range(1, limit + 1)}
    primes[1] = False
    # primes = [True] * (limit + 1)
    p = 2
    while p ** 2 <= limit:
        if primes[p]:
            for i in range(p ** 2, limit + 1, p):
                primes[i] = False
        p += 1
    return primes


def goldbach_partitions_count(even_number, primes):
    count = 0
    for prime in primes:
        if not primes[prime]:
            continue
        if prime > even_number // 2:
            break
        if primes[even_number - prime]:
            count += 1
    return count


limit = 1000000  # 충분히 큰 범위의 소수 리스트 생성
primes = sieve_of_eratosthenes(limit)
print([p for p in range(1, limit + 1) if primes[p] == True])
n = int(input())
for _ in range(n):
    even_number = int(input())
    partitions_count = goldbach_partitions_count(even_number, primes)
    print(partitions_count)

File: 2023_8/baekjoon/17103.rs
use std::io::{stdin, Read};

fn main() {
    let mut buf = String::new();
    stdin().read_to_string(&mut buf).unwrap();
    let mut buf = buf.split_ascii_whitespace();
    let mut iterator = || buf.next().unwrap().parse::<usize>().unwrap();
    let primes = get_primes();

    for _ in 0..iterator() {
        let mut result = 0;
        let n = iterator();
        for prime in 1..(1_000_000 + 1) {
            if !primes[prime] {
                continue;
            }
            if prime > (n / 2) {
                break;
            }
            if primes[n - prime] {
                result += 1;
            }
        }
        println!("{}", result);
    }
}

fn get_primes() -> Vec<bool> {
    let limit = 1000000;
    let mut nums = vec![true; limit + 1];
    nums[1] = false;

    let mut i = 2;
    while i * i <= limit {
        if nums[i] {
            for j in ((i * i)..(limit + 1)).step_by(i) {
                nums[j] = false;
            }
        }
        i += 1;
    }

    nums
}
